Content.Shared/Containers/ItemSlot/ItemSlotsSystem.cs

```diff
@@ -199,20 +207,45 @@ private void OnInteractUsing(EntityUid uid, ItemSlotsComponent itemSlots, Intera
             if (!EntityManager.TryGetComponent(args.User, out HandsComponent? hands))
                 return;
 
+            if (itemSlots.Slots.Count == 0)
+                return;
+
             var slots = new List<ItemSlot>();
+
+            var allLocked = false;
+            ItemSlot? lastFailedSlot = null;
             foreach (var slot in itemSlots.Slots.Values)
             {
                 if (!slot.InsertOnInteract)
                     continue;
 
-                if (!CanInsert(uid, args.Used, args.User, slot, swap: slot.Swap, popup: args.User))
+                if (!CanInsert(uid, args.Used, args.User, slot, slot.Swap))
+                {
+                    if (!slot.Locked)
+                        allLocked = false;
+                    lastFailedSlot = slot;
                     continue;
+                }
 
                 slots.Add(slot);
             }
 
             if (slots.Count == 0)
+            {
+                // no slots failed, so uh,
+                if (lastFailedSlot == null)
+                    return;
+
+                // it's a bit weird that the popupMessage is stored with the item slots themselves, but in practice
+                // the popup messages will just all be the same, so it's probably fine.
+                //
+                // doing a check to make sure that they're all the same or something is probably frivolous
+                if (lastFailedSlot.WhitelistFailPopup != null)
+                    _popupSystem.PopupClient(Loc.GetString(lastFailedSlot.WhitelistFailPopup), uid, args.User);
+                else if (allLocked && lastFailedSlot.LockedFailPopup != null)
+                    _popupSystem.PopupClient(Loc.GetString(lastFailedSlot.LockedFailPopup), uid, args.User);
```

Comment:
Does this properly show the WhitelistFailPopup if not all slots are locked but the last slot is not whitelisted?

Reply:
Wait, I just realized you'd want to show the locked popup if there's a slot that the item could go in, but the slot is locked. 

OKAY, so the logic goes like this:

*   `lastValidLockedSlot` stores, if applicable, the last slot that should technically allow the item but is locked
*   go through every single valid interactable slot
*   if you can't insert the item,
    *   if the item is allowed by the whitelist / blacklist, but the slot is locked, set `lastValidLockedSlot` 
*   if there are no valid slots, 
    *   if `lastValidLockedSlot` is set, display the `lockedFailPopup`
    *   otherwise, display the `whitelistFailPopup`
*   otherwise, if there are valid slots
    *   insert the item into the valid slot as normal

If none of the slots would have let the item in, it will display the `whitelistFailPopup`

If there was at least slot that (maybe) would have let it in if it wasnt locked, and there are no other valid slots, display `lockedFailPopup`.

I think it's correct now. I can't test it in game because the only machine that takes advantage of the `lockedFailPopup` is the AME.

I did test it again on the reagent dispenser, and it seems to work so.